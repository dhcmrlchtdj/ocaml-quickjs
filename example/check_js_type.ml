let ( let* ) = Stdlib.Result.bind

let ok_unit = Ok ()

let _ =
  let open Quickjs.Value in
  let _ =
    let ctx = Quickjs.new_runtime () |> Quickjs.new_context in
    let* d1 = Quickjs.eval ~ctx "new Date()" in
    let* d2 = Quickjs.eval ~ctx "Date" in
    if is_instance_of d1 d2
    then print_endline "instanceof | pass"
    else print_endline "instanceof | fail";
    ok_unit
  in
  let _ =
    let* x = Quickjs.eval "1" in
    let y = convert_to_string x in
    if is_number x && is_string y
    then print_endline "convert_to_string | pass"
    else print_endline "convert_to_string | fail";
    ok_unit
  in
  let _ =
    let cases =
      [
        (is_null, "null");
        (is_undefined, "undefined");
        (is_bool, "true");
        (is_number, "1");
        (is_string, {|"string"|});
        (is_symbol, "Symbol()");
        (is_array, "[]");
        (is_object, "({obj:true})");
        (is_function, "parseInt");
        (is_constructor, "String");
        (is_error, "new Error()");
        (is_big_int, "1n");
        (is_big_decimal, "1m");
        (is_big_float, "1l");
      ]
    in
    let iter_f (f, s) =
      let v = Quickjs.eval_unsafe s in
      let r = if f v then s ^ " | pass" else s ^ " | fail" in
      print_endline r
    in
    List.iter iter_f cases
  in
  ()
