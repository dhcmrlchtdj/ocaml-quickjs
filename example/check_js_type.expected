instanceof | pass
convert_to_string | pass
null | pass
undefined | pass
true | pass
1 | pass
"string" | pass
Symbol() | pass
[] | pass
({obj:true}) | pass
parseInt | pass
String | pass
new Error() | pass
1n | pass
1m | pass
1l | pass
